cpp/src/arrow/adapters/orc/adapter.h

```diff
@@ -0,0 +1,65 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#ifndef ARROW_ORC_CONVERTER_H
+#define ARROW_ORC_CONVERTER_H
+
+#include <cstdint>
+#include <list>
+#include <memory>
+
+#include "arrow/io/interfaces.h"
+#include "arrow/memory_pool.h"
+#include "arrow/record_batch.h"
+#include "arrow/status.h"
+#include "arrow/type.h"
+#include "arrow/util/visibility.h"
+
+namespace arrow {
+
+namespace adapters {
+
+namespace orc {
+
+class ARROW_EXPORT ORCFileReader {
+ public:
+  ~ORCFileReader();
+  static Status Open(const std::shared_ptr<io::ReadableFileInterface>& file,
+                     MemoryPool* pool, std::unique_ptr<ORCFileReader>* reader);
+  Status ReadSchema(std::shared_ptr<Schema>* out);
+  Status Read(std::shared_ptr<RecordBatch>* out);
+  Status Read(const std::list<uint64_t>& include_indices,
+              std::shared_ptr<RecordBatch>* out);
```

Comment:
This reads an entire file as a single `RecordBatch`. I'm not sure if this is best, or if it should allocate a bunch of smaller `RecordBatch`'s and return a `Table` composed of them.

I'm also not sure if the api here should be working with `Table` instead of `RecordBatch`.

Reply:
Reading the full file should go into a `Table` instance. If we are reading only chunks of a file, they should than go into a `RecordBatch` which then could be merged by the user into a `Table`.

Reply:
Should the resulting table be composed of a single or multiple record batches? Either is doable, as we're having to copy data out of the orc reader in either case. Currently I read the whole table as a single large record batch, but there may be some benefit to breaking it into a collections of smaller blocks, not sure.

Reply:
@wesm, @cpcloud any strong thoughts on this ^^?

Reply:
The problem with this is that if there is a string column with a lot of very large strings, you could overrun the capacity of `arrow::BinaryArray` -- so it would be better to yield a Table even if all the columns are not chunked

I missed this in my review, but it would be better to use `std::vector<int>` here, a bit more consistent with the Parquet reader

Reply:
Ok, I've changed the `Read` method to return a table composed of one record batch per stripe. I'm not sure if this is the best api - could also see returning a table composed of fixed-length record batches given a user proved batch-size (which would be equally efficient, since we have to copy the data either way).

Reply:
> I missed this in my review, but it would be better to use std::vector<int> here, a bit more consistent with the Parquet reader

Done. One question I have now is when to use `int64_t` and when to just use `int`. Right now the user facing api uses a mix, which seems off. Not sure what the best practice is here, I see both used in the arrow codebase.

Reply:
^^^ Any response to this?